app/assets/javascripts/initializers/initializeDisplayAdVisibility.js

```diff
@@ -14,6 +14,110 @@ function initializeDisplayAdVisibility() {
       ad.classList.remove('hidden');
     }
   });
+}
+
+function trackAdImpression(adBox) {
+  var isBot = /bot|google|baidu|bing|msn|duckduckbot|teoma|slurp|yandex/i.test(
+    navigator.userAgent,
+  ); // is crawler
+  var adSeen = adBox.dataset.impressionRecorded;
+  if (isBot || adSeen) {
+    return;
+  }
+
+  var tokenMeta = document.querySelector("meta[name='csrf-token']");
+  var csrfToken = tokenMeta && tokenMeta.getAttribute('content');
+
+  var dataBody = {
+    display_ad_event: {
+      display_ad_id: adBox.dataset.id,
+      context_type: adBox.dataset.contextType,
+      category: adBox.dataset.categoryImpression,
+    },
+  };
 
-  // for impression & click-tracking, see initializeBaseTracking
+  window
+    .fetch('/display_ad_events', {
+      method: 'POST',
+      headers: {
+        'X-CSRF-Token': csrfToken,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(dataBody),
+      credentials: 'same-origin',
+    })
+    .catch((error) => console.error(error));
```

Comment:
Suggestion: consider extracting this fetch/catch to a helper function

Reply:
Another instance where I've [extracted this from elsewhere](https://github.com/forem/forem/blob/2b23d4f6ed0bd01119854a92b36ec2036282d0d3/app/assets/javascripts/initializers/initializeBaseTracking.js.erb#L241-L251) - and I think that refactor can happen in a later pass.